CLOUDP-282641: Use oauth2 token source

wtrocki opened the pull request:
## Description

Change to implementation to reuse oauth2 token source internally to perform token requests.
For discussion/spec see: CLOUDP-282641

## Progress

- [x] Updates to the codebase
- [x] Unit tests
- [x] Examples 
- [x] Documentation 
- [x] Custom User Agent

wtrocki commented on auth/credentials/api.go:
```diff
@@ -63,40 +65,58 @@ func NewTokenSourceWithOptions(opts AtlasTokenSourceOptions) TokenSource {
 
-	if opts.TokenCache == nil {
-		opts.TokenCache = &InMemoryTokenCache{}
+	config := &clientcredentials.Config{
```
Using oauth2 config to create oauth2.TokenSource

wtrocki commented on auth/credentials/api.go:
```diff
@@ -63,40 +65,58 @@ func NewTokenSourceWithOptions(opts AtlasTokenSourceOptions) TokenSource {
 	}
 }
+
+// RevokeToken checks if TokenSource is a RevocableTokenSource and revokes Token.
```
Since `oauth2.TokenSource` does not have revocation, we're casting `RevocableTokenSource`

wtrocki replied:
Actually. I will push a commit that will use RevocableTokenSource as internal TokenSource. We can always revert it.

wtrocki commented on auth/credentials/oauth.go:
```diff
@@ -66,142 +64,82 @@ func (c *OAuthTokenSource) RevokeToken() error {
 	}, nil
 }
 
-// fetchTokenFromRemoteServer makes a manual POST request to Server (tokenUrl) to fetch the access Token.
```
This is currently done by oauth2 lib. 
With this change we're missing the ability to set User-Agent and control error handling, retry logic etc.

gssbzn commented on auth/credentials/oauth.go:
```diff
@@ -11,41 +11,39 @@ import (
-}
 
-// TokenSource interface allows to fetch and revoke valid OAuth Access Tokens.
-type TokenSource interface {
```
for simplicity you can do 
```go 
type TokenSource = oauth2.TokenSource
```

wtrocki replied:
SGTM. TokenSource is not exposed externally (only used internally). 
This line removes it. What will be the purpose of redefining token source in this case?

gssbzn replied:
it is, it's a public package with an exported name, if you really don't want people to use these types or methods then make them all unexported or `/internal` and we can move on

wtrocki replied:
We are exporting `RevocableTokenSource` in api.go. oauth2.Token source is only used as interface we inherit from.
That line removes TokenSource. Good point on not exposing only needed API. 
I will make oAuthTokenSource private as it is not needed to be exposed directly.

wtrocki commented on auth/credentials/api.go:
```diff
@@ -63,40 +65,52 @@ func NewTokenSourceWithOptions(opts AtlasTokenSourceOptions) TokenSource {
+		ClientID:     opts.ClientID,
+		ClientSecret: opts.ClientSecret,
+		TokenURL:     tokenURL,
+		AuthStyle:    oauth2.AuthStyleInHeader,
+		Scopes:       []string{}, // No support for scopes.
```
I prefer to flag `// No support for scopes.` in the wrapper to avoid potential help ticket asking to expose that in admin sdk. Happy to remove it.

lantoli commented on auth/credentials/oauth_test.go:
```diff
@@ -167,12 +174,10 @@ func TestOAuthClient_FetchToken_Failure(t *testing.T) {
-	oAuthTokenSource := tokenSource.(*OAuthTokenSource)
-	assert.Equal(t, mockServer.URL+tokenAPIPath, oAuthTokenSource.tokenURL)
-	assert.Contains(t, err.Error(), "Failed to obtain Access Token when fetching new OAuth Token from remote server")
+	assert.Contains(t, err.Error(), "oauth2: cannot fetch token: 401 Unauthorized")
```
nit: you can use assert.ErrorContains

wtrocki replied:
Nice! Yes. Good protip.

lantoli commented on auth/credentials/oauth_test.go:
```diff
@@ -218,10 +223,10 @@ func TestOAuthClient_FetchToken_FailureHtmlContentType(t *testing.T) {
+	assert.Contains(t, err.Error(), "oauth2: cannot fetch token: 403 Forbidden")
 }
 
 // Test CustomTransport to ensure Token is injected into requests
```
not changed in this PR, but prefer t.Cleanup to defer in tests, e.g.: https://brandur.org/fragments/go-prefer-t-cleanup-with-parallel-subtests

lantoli commented on go.sum:
```diff
@@ -10,6 +10,8 @@ github.com/stretchr/objx v0.5.2 h1:xuMeJ0Sdp5ZMRXx/aWO6RZxdr3beISkG5/G/aIRr3pY=
 github.com/stretchr/testify v1.9.0 h1:HtqpIVDClZ4nwg75+f6Lvsy/wHu+3BoSGCbBAcpTsTg=
 github.com/stretchr/testify v1.9.0/go.mod h1:r2ic/lqez/lEtzL7wO/rwa5dbSLXVDPFyf8C91i36aY=
+golang.org/x/oauth2 v0.23.0 h1:PbgcYx2W7i4LvjJWEbf0ngHV6qJYr86PkAV3bXdLEbs=
+golang.org/x/oauth2 v0.23.0/go.mod h1:XYTD2NtWslqkgxebSiOHnXEap4TF09sJSc7H1sXbhtI=
```
should we say something in the doc that we depend on this module?

wtrocki replied:
We do not want to expose that information. We wrap all oauth2 fields.

gssbzn commented on auth/credentials/api.go:
```diff
@@ -22,81 +19,22 @@ const serverTokenURL = core.DefaultCloudURL + tokenAPIPath
-			TokenSource:         client,
-		},
-	}
+//func NewTokenSource(clientID, clientSecret string) RevocableTokenSource {
```
not sure if this still needed

wtrocki replied:
Not needed. Current approach using headless TokenSource (We create new TokenSource every time token is requested and use PersistentStore to reuse TokenSources.

wtrocki replied:
Removed

gssbzn commented on admin/atlas_client.go:
```diff
@@ -58,24 +60,11 @@ func UseDigestAuth(apiKey, apiSecret string) ClientModifier {
 // Warning: for advanced use cases please use credentials.NewTokenSource directly in your code pass it to UseHTTPClient method.
 // Warning: any previously set httpClient will be overwritten. To fully customize HttpClient use UseHTTPClient method.
-func UseOAuthAuth(clientID, clientSecret string, tokenCache credentials.LocalTokenCache) ClientModifier {
+func UseOAuthAuth(clientID, clientSecret string, tokenCache *oauth2.Token) ClientModifier {
```
this should probably accept a context and not sure if other modifiers like the ones for the base url

wtrocki replied:
This concept was inherited from old SDK, and it does hold us back due to `UseBaseURL` or `UseHttpClient` overriding effects of this one.

gssbzn replied:
I'm not talking about the method but that this method should probably accept a ctx and should maybe allow to override the base url, of the oauth client which right now it doesn't allow me and why I could not really test this

wtrocki replied:
UseBaseURL overrides URL global. That is why `&c.Servers[0].URL` were used in the code that was removed.
Context should be also reused across the entire client otherwise we will have very confusing end user API.

wtrocki replied:
Adding those changes.

gssbzn replied:
>UseBaseURL overrides URL global. That is why &c.Servers[0].URL were used in the code that was removed.

Does that mean I need to know I have to call that function first?

wtrocki replied:
Yes. That is why I think we should:
- Deprecate UseDigest 
- Change examples to UseHttpClient instead (for both digest and OAuth)

WDYT?

gssbzn replied:
no strong opinion, the CLI already works with `UseHttpClient`

wtrocki replied:
I think with this PR we moving from:

- SDK provides simple and comprehensive 
- Just use oauth2 + our solution for token storage we provided.

gssbzn replied:
yes, which should've always been the case :)

gssbzn commented:
if you don't need to load a cache token
```go
package main

import (
	"context"
	"fmt"
	"time"

	"log"

	atlas "go.mongodb.org/atlas-sdk/v20241023001/admin"
	"go.mongodb.org/atlas-sdk/v20241023001/auth/credentials"
)

func main() {
	conf := credentials.Config{
		RevokeURL: "http://localhost:8080/api/oauth/revoke",
	}
	conf.ClientID = "mdb_sa_id_<>"
	conf.ClientSecret = "mdb_sa_sk_<>"
	conf.TokenURL = "http://localhost:8080/api/oauth/token"
	ctx := context.Background()
	// t, err := conf.TokenSource(ctx).Token()
	// if err != nil {
	// 	log.Fatalln(err.Error())
	// }

	tc := conf.Client(ctx)

	client, err := atlas.NewClient(atlas.UseHTTPClient(tc), atlas.UseBaseURL("http://localhost:8080/"))
	if err != nil {
		log.Fatalln(err.Error())
	}
	projects, _, err := client.ProjectsApi.ListProjects(ctx).Execute()
	if err != nil {
		log.Fatalln(err.Error())
	}

	for _, p := range projects.GetResults() {
		time.Sleep(5 * time.Second)
		fmt.Println("projectID", p.GetId())
		clusters, _, err := client.ClustersApi.ListClusters(ctx, p.GetId()).Execute()
		if err != nil {
			log.Fatalln(err.Error())
		}
		for _, c := range clusters.GetResults() {
			fmt.Println("clusterName", c.GetName())
		}
	}
	err = conf.RevokeToken(ctx, t)
	if err != nil {
		log.Fatalln(err.Error())
	}
}

```

wtrocki commented on admin/atlas_client.go:
```diff
@@ -58,24 +60,11 @@ func UseDigestAuth(apiKey, apiSecret string) ClientModifier {
-		}
-		httpClient := credentials.NewHTTPClientWithOAuthToken(tokenSource)
+		oauth := credentials.NewConfig(clientID, clientSecret)
+		source := oauth2.ReuseTokenSource(tokenCache, oauth.TokenSource(context.Background()))
```
[Q] Why using `ReuseTokenSource` from oauth2.

gssbzn replied:
because it has the logic to check if the token is valid and reuse it or request a new one and keep using it while valid, I don't think we need to implement that logic ourselves

wtrocki replied:
Prior to your commit, wrapper of ReuseToken source provided the ability to save OAuth token.

gssbzn replied:
yes for file systems, for in memory there's nothing to save, it just works

wtrocki replied:
Added method for saving token in filesystem

wtrocki commented on auth/credentials/clientcredentials.go:
```diff
@@ -0,0 +1,78 @@
+// Config describes a 2-legged OAuth2 flow, with both the
+// client application information and the server's endpoint URLs.
+type Config struct {
+	clientcredentials.Config
```
Do you have strong opinion on using config?
I'm concerned that this produces end user api that will expose too much.
Some examples:
```
	        // Token URL should be inherited from SDK. For all users will be hardcoded and not exposed to users.
		TokenURL:     "https://cloud.mongodb.com/oauth/token",
		// Scopes are present but they are not supported by our backend
		Scopes:       []string{"your-scope"},
            // By default will make redundant wrong requests against our API.
	      AuthStyle: oauth2.Auto
```

gssbzn replied:
yes it has the implementation to get tokens, if we don't use it you have to implement them which was the whole point of this change, not re-implement oauth

to your concerns there's NewConfig which encapsulates this logic to create the one you want, so following your comments of "users won't use this package directly, then they won't use this struct directly right?

on top of that for this one 
>// By default will make redundant wrong requests against our API.

on my tests I haven't' seen this to be true

wtrocki replied:
> if we don't use it you have to implement them which was the whole point of this change, not re-implement oauth

Prior to last commit, all changes we had reused `oauth.TokenSource.Token()` but also incorporated additional requirements from the scope: 
- Ability to integrate with Atlas CLI by creating transport instead of Client.
- Token being string. 
- Custom user agent etc. 
- RevokableTokenSource that provided immutable access to Token and Revoke so it can be part of the context without security risks.   

I like how lean and clean current approach is but I might need to reintroduce some removed code in order to support those features. Is that ok?

> on my tests I haven't' seen this to be true

We have provided override when creating config, but we are still exposing that field to end users which is my point.

gssbzn replied:
I don't see any of the points you mention captured in the scope

wtrocki replied:
Covered offline. We will settle with oauth2.Token as representation of token. Apply custom user agent.
We do not need RevocableTokenSource as plan is to contribute those changes to auth2 in long run.

wtrocki commented on auth/credentials/clientcredentials.go:
```diff
@@ -0,0 +1,78 @@
+	userAgent string
+}
+
+// tokenSource manages the OAuth Token fetching and refreshing using a LocalTokenCache.
```
`LocalTokenCache` seems to have been removed?

gssbzn replied:
correct

wtrocki commented on auth/credentials/api.go:
```diff
@@ -22,81 +19,22 @@ const serverTokenURL = core.DefaultCloudURL + tokenAPIPath
+	c := &Config{}
+	c.ClientID = clientID
+	c.ClientSecret = clientSecret
+	c.RevokeURL = revokeAPIPath
```
How this is supported by Token library?

wtrocki commented on auth/credentials/api.go:
```diff
@@ -22,81 +19,22 @@ const serverTokenURL = core.DefaultCloudURL + tokenAPIPath
+	c := &Config{}
+	c.ClientID = clientID
+	c.ClientSecret = clientSecret
+	c.RevokeURL = revokeAPIPath
```
Should I use path subtraction or replace this to be whole url?

gssbzn replied:
yeah bug, fixing

wtrocki commented on auth/credentials/api.go:
```diff
@@ -22,81 +19,22 @@ const serverTokenURL = core.DefaultCloudURL + tokenAPIPath
+//	})
+//}
+
+func NewConfig(clientID, clientSecret string) *Config {
```
Any reason we need to return oauth2.Config as end user API. Could we create token source here?

gssbzn replied:
we are returning a wrapper for `oauth2.Config` not `oauth2.Config` directly, the reason is that revoking is not part of TokenSource is part of Config and if you return TokenSource you loose access to revoke

wtrocki commented on auth/credentials/clientcredentials.go:
```diff
@@ -0,0 +1,78 @@
+}
+
+func (c *tokenSource) Token() (*oauth2.Token, error) {
+	return c.conf.Token(c.ctx)
```
[Question] MIssing info how user agent is passed to token source.

wtrocki replied:
Decision. We will bring previous change to supply user agent properly.

wtrocki commented on auth/credentials/clientcredentials.go:
```diff
@@ -28,7 +29,7 @@ type tokenSource struct {
 // RevokeToken revokes the Access Token while also removing it from the Access Token Cache.
 // When the Access Token is expired or missing, revoke will return without any action.
-func (c *Config) RevokeToken(ctx context.Context, t *oauth2.Token) error {
+func (c *Config) RevokeToken(ctx context.Context, t *auth.Token) error {
```
We removed logic that accepted token string. 
To pass token we will require client to store expiration or parse token into auth.Token.

gssbzn replied:
yes, I'm not a fan of the previous approach parsing the string to guess the expiry date, specially when that parsing wasn't doing any sort of signature validation of the token so they could easily be tampered strings, I strongly belive that we shouldn't be dealing with strings but a Token object, this is similar to what the official oauth lib does

wtrocki replied:
That means that customer applications will need to be dealing with building Token (from storage etc.)
Tampered strings/Token do not matter as all oauth backend does provide security features in this case. 
Client side uses expiration to avoid sending expired tokens only.

gssbzn replied:
the token type is (de)serializable, it'll be safer to operate within the struct than string parsing of the token and risk a tampered one, even if the back-end would still catch it

wtrocki replied:
SGTM

wtrocki commented:
@gssbzn Thank you for changes. I will take from there. Might need to reintroduce some of the code to fill the gaps and request your review.

wtrocki commented on auth/credentials/transport.go:
```diff
@@ -31,6 +32,27 @@ func (t *OAuthCustomHTTPTransport) RoundTrip(req *http.Request) (*http.Response,
+}
+
+func (t *userAgentTransport) RoundTrip(req *http.Request) (*http.Response, error) {
+	req.Header.Set("User-Agent", t.userAgent) // Set the custom User-Agent
```
Done

wtrocki commented:
## Update

1. Included all changes by @gssbzn 
2. Added way to send user agent when requesting token and store token when it is refreshed
3. Updated examples and unit tests.

gssbzn commented on admin/atlas_client.go:
```diff
@@ -55,27 +57,35 @@ func UseDigestAuth(apiKey, apiSecret string) ClientModifier {
-				TokenCache:   tokenCache,
-				BaseURL:      &c.Servers[0].URL,
+		oauth := credentials.NewConfig(clientID, clientSecret)
+		ctx := context.WithValue(context.Background(), oauth2.HTTPClient,
```
should we be passing the context as and argument?

wtrocki replied:
Yes. Makes sense.

gssbzn commented on admin/atlas_client.go:
```diff
@@ -55,27 +57,35 @@ func UseDigestAuth(apiKey, apiSecret string) ClientModifier {
+
+// UseOAuthAuthWithCache provides OAuthAuth authentication for Go SDK with the ability to reuse cached OAuth Token
+// Method is provided as helper to create a default HTTP client that supports OAuth (Service Accounts) authentication.
+// Warning: any previously set httpClient will be overwritten. To fully customize HttpClient use UseHTTPClient method.
```
just becuase is in a new line doesn't mean it renders in a new line 
```suggestion
//
// Warning: any previously set httpClient will be overwritten. To fully customize HttpClient use UseHTTPClient method.
```

gssbzn commented on auth/credentials/clientcredentials.go:
```diff
@@ -0,0 +1,69 @@
+
+// Config describes a 2-legged OAuth2 flow, with both the
+// client application information and the server's endpoint URLs.
+// NOTE: clientcredentials.Config values are used only internally
```
```suggestion
//
// NOTE: clientcredentials.Config values are used only internally
```

wtrocki commented on auth/credentials/transport.go:
```diff
@@ -1,36 +1,38 @@
-func (t *OAuthCustomHTTPTransport) setAuthHeader(r *http.Request, token *Token) {
-	r.Header.Set("Authorization", "Bearer "+token.AccessToken)
+// NewOAuthCacheTransport http client that adds the User-Agent header and TokenCache
+func NewOAuthCacheTransport(transport http.RoundTripper, userAgent *string, tokenCache *func(token string) error) http.RoundTripper {
```
Yes. I have tried to make it part of the admin as It is used in admin to enable admin package specific features we need. 
However admin package is fully generated.

wtrocki commented on auth/credentials/transport.go:
```diff
@@ -1,36 +1,38 @@
-	// TokenSource used to obtain valid OAuth Token
-	TokenSource TokenSource
+// OAuthTransport provides a way to cache OAuth Token
+type OAuthTransport struct {
```
Will make this private.

wtrocki commented:
Applied fixes. I will add Token cache example in separate PR.

wtrocki reviewed (commented):
@gssbzn changes LGTM

jeroenvervaeke commented on admin/atlas_client.go:
```diff
@@ -55,28 +57,24 @@ func UseDigestAuth(apiKey, apiSecret string) ClientModifier {
 
 // UseOAuthAuth provides OAuthAuth authentication for Go SDK.
```
```suggestion
// UseOAuthAuth provides OAuth authentication for Go SDK.
```

jeroenvervaeke reviewed (approved):
One small comment about docs, besides that LGTM
